roslibrust_rospack/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+    pub path: PathBuf,
+    pub is_metapackage: bool,
+}
+
+const CATKIN_IGNORE: &'static str = "CATKIN_IGNORE";
+const PACKAGE_FILE_NAME: &'static str = "package.xml";
+const ROS_PACKAGE_PATH_ENV_VAR: &'static str = "ROS_PACKAGE_PATH";
+
+pub fn get_search_paths() -> Vec<PathBuf> {
+    if let Ok(paths) = std::env::var(ROS_PACKAGE_PATH_ENV_VAR) {
+        #[cfg(unix)]
+        let separator = ":";
+        #[cfg(windows)]
+        let separator = ";";
+
+        paths
+            .split(separator)
+            .map(|path| PathBuf::from(path))
+            .collect::<Vec<PathBuf>>()
+    } else {
+        eprintln!("No ROS_PACKAGE_PATH defined.");
+        vec![]
+    }
+}
+
+pub fn crawl(search_paths: Vec<PathBuf>) -> Vec<Package> {
+    let mut packages = vec![];
+
+    for path in search_paths {
+        if let Ok(found_packages) = packages_from_path(path, 1000) {
+            packages = [packages, found_packages].concat();
+        }
+    }
+
+    packages
+}
+
+fn packages_from_path(mut path: PathBuf, depth: u16) -> io::Result<Vec<Package>> {
+    let mut found_packages = vec![];
+
+    if depth == 0 {
+        eprintln!(
+            "Reached depth limit in: {}",
+            path.as_os_str().to_string_lossy()
+        );
+        return Err(io::ErrorKind::Other.into());
+    }
+
+    if path.as_path().is_dir() {
+        // We have a valid path
+        path.push(CATKIN_IGNORE);
+        // We'll only check this directory if no CATKIN_IGNORE file is present
+        if !path.as_path().is_file() {
+            assert!(path.pop());
+
+            path.push(PACKAGE_FILE_NAME);
+            if path.as_path().is_file() {
+                // And there's a package.xml here!
+                let package_xml_path = path.clone();
+                assert!(path.pop());
+                found_packages.push(Package {
+                    name: String::from(path.file_name().unwrap().to_string_lossy()),
+                    path: path,
+                    is_metapackage: is_metapackage(package_xml_path.as_path()).unwrap_or(false),
+                });
+            } else {
+                // No file here, we'll have to go deeper
+                assert!(path.pop());
+                for subdir in std::fs::read_dir(path)
+                    .unwrap()
+                    .filter(|entry| match entry {
+                        Ok(entry) => entry.path().as_path().is_dir(),
+                        Err(_err) => false,
+                    })
+                    .map(|entry| entry.unwrap())
+                {
+                    found_packages = [
+                        found_packages,
+                        packages_from_path(subdir.path(), depth - 1)?,
+                    ]
+                    .concat()
+                }
+            }
+        }
+    } else {
+        eprintln!("{} is not a directory", path.to_string_lossy())
+    }
+
+    Ok(found_packages)
+}
+
+fn is_metapackage(package_xml_path: &Path) -> io::Result<bool> {
```

Comment:
Is there a test for this? Not super familiar with these but seems slightly wonky that we would need to recognize

Reply:
I believe I added this because it broke my initial implementation, but I can't actually find any examples that would cause that (and the spec for metapackages seems to indicate they shouldn't). I'll remove all the metapackage stuff.